### PR TITLE
Refactor: Reorganize code for Forecasting and AnomalyDetector packages

### PR DESCRIPTION
**Pull request by kaituo**
### Description
This PR includes several refactorings to increase code reuse and improve readability:

1. Entity, FeatureData, and RestHandlerUtils have been moved from the 'ad' package to the 'timeseries' package, enabling their use within the Forecasting code.
2. A new class, DataByFeatureId, has been created to hold shared code previously located in FeatureData. FeatureData now inherits from DataByFeatureId. The standalone usage of DataByFeatureId will be detailed in subsequent PRs.
3. Renamed checkAnomalyDetectorFeaturesSyntax to checkFeaturesSyntax in RestHandlerUtils to make it more generic and usable for Forecasting.
4. Constants AGG_NAME_MAX_TIME, AGG_NAME_MIN_TIME, DATE_HISTOGRAM, and FEATURE_AGGS have been moved from ADCommonName to CommonName to make them accessible for Forecasting.
5. A new method, parseConfig, has been added to the Config class. This method can parse configuration for either AnomalyDetector or Forecaster based on input.

Testing:
- The changes have been validated with a successful gradle build.

### Issues Resolved
https://github.com/opensearch-project/anomaly-detection/issues/866

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/timeseries/model/Config.java [hunk 1]
```diff
@@ -559,4 +562,14 @@ protected void checkAndThrowValidationErrors(ValidationAspect validationAspect)
             throw new TimeSeriesException(CommonMessages.FAIL_TO_VALIDATE);
         }
     }
+
+    public static Config parseConfig(Class<? extends Config> configClass, XContentParser parser) throws IOException {
```
Comment (amitgalitz) [inline src/main/java/org/opensearch/timeseries/model/Config.java:566]: I see this hasn't been used yet, I assume you'll use it later on?

Comment (kaituo) [reply src/main/java/org/opensearch/timeseries/model/Config.java:566]: yes

File: src/main/java/org/opensearch/timeseries/util/RestHandlerUtils.java [hunk 2]
```diff
@@ -130,18 +135,18 @@ public static XContentParser createXContentParserFromRegistry(NamedXContentRegis
...
         if (features != null) {
-            if (features.size() > maxAnomalyFeatures) {
-                return "Can't create more than " + maxAnomalyFeatures + " anomaly features";
+            if (features.size() > maxFeatures) {
+                return "Can't create more than " + maxFeatures + " features";
             }
-            return validateFeaturesConfig(anomalyDetector.getFeatureAttributes());
+            return validateFeaturesConfig(config.getFeatureAttributes());
         }
         return null;
```
Comment (amitgalitz) [inline src/main/java/org/opensearch/timeseries/util/RestHandlerUtils.java:151]: Also in the method `validateFeaturesConfig` in this file we have this line: `errorMsgBuilder.append("Detector has duplicate feature names: ");` We should probably change to either "There are duplicate feature names"/"Given configurations have duplicate feature names" or exact the class being used (Forecaster/detector) for the given messaging. In short term wont really be an issue since max feature number will be one for forecasting

File: src/main/java/org/opensearch/timeseries/util/RestHandlerUtils.java [hunk 3]
```diff
@@ -130,18 +135,18 @@ public static XContentParser createXContentParserFromRegistry(NamedXContentRegis
...
     }
```
[the remark this answers is not in the document]
Comment (kaituo) [reply src/main/java/org/opensearch/timeseries/util/RestHandlerUtils.java:152]: Yes. Changed comment on line 200 like

```
/**
     * Wrap action listener to avoid return verbose error message and wrong 500 error to user.
     * Suggestion for exception handling in timeseries analysis (e.g., AD and Forecast):
     * 1. If the error is caused by wrong input, throw IllegalArgumentException exception.
     * 2. For other errors, please use TimeSeriesException or its subclass, or use
     *    OpenSearchStatusException.
     *
     * TODO: tune this function for wrapped exception, return root exception error message
     *
     * @param actionListener action listener
     * @param generalErrorMessage general error message
     * @param <T> action listener response type
     * @return wrapped action listener
     */
    public static <T> ActionListener wrapRestActionListener(ActionListener<T> actionListener, String generalErrorMessage) {
        return ActionListener.<T>wrap(r -> { actionListener.onResponse(r); }, e -> {
            logger.error("Wrap exception before sending back to user", e);
            Throwable cause = Throwables.getRootCause(e);
            if (isProperExceptionToReturn(e)) {
                actionListener.onFailure(e);
            } else if (isProperExceptionToReturn(cause)) {
                actionListener.onFailure((Exception) cause);
            } else {
                RestStatus status = isBadRequest(e) ? BAD_REQUEST : INTERNAL_SERVER_ERROR;
                String errorMessage = generalErrorMessage;
                if (isBadRequest(e) || e instanceof TimeSeriesException) {
                    errorMessage = e.getMessage();
                } else if (cause != null && (isBadRequest(cause) || cause instanceof TimeSeriesException)) {
                    errorMessage = cause.getMessage();
                }
                actionListener.onFailure(new OpenSearchStatusException(errorMessage, status));
            }
        });
    }
```

File: src/main/java/org/opensearch/timeseries/model/Entity.java [hunk 4]
```diff
@@ -9,7 +9,7 @@
  * GitHub history for details.
  */
 
-package org.opensearch.ad.model;
+package org.opensearch.timeseries.model;
```
Comment (amitgalitz) [inline src/main/java/org/opensearch/timeseries/model/Entity.java:12]: nit: we use detectorId in some method input name which should change now to configID or timeSeriesID. Also some mentions of detector in comments (these comments though also refer to bwc between versions that are only really relatable to AD and are helpful as they are though so I don't mind to either keep as is (e.g. getModelId) and maybe just mentioning nuance when applied to new forecaster. Can also be dealt with when we get to the point forecasting code is calling these methods.

Comment (kaituo) [reply src/main/java/org/opensearch/timeseries/model/Entity.java:12]: I fixed comments and parameters of getModelId. I also modified getModelId to not consider bwc when this is used in forecasting.

File: src/main/java/org/opensearch/timeseries/model/Config.java [hunk 5]
```diff
@@ -559,4 +562,14 @@ protected void checkAndThrowValidationErrors(ValidationAspect validationAspect)
...
+        if (configClass == AnomalyDetector.class) {
+            return AnomalyDetector.parse(parser);
+        } else if (configClass == Forecaster.class) {
+            return Forecaster.parse(parser);
+        } else {
+            throw new IllegalArgumentException("Unsupported config type");
```
Comment (owaiskazi19) [inline src/main/java/org/opensearch/timeseries/model/Config.java:572]:
```suggestion
            throw new IllegalArgumentException("Unsupported config type. Supported config types are [AnomalyDetector, Forecaster]");
```

**Review by owaiskazi19 — COMMENTED** [review]
Minor comments. Overall LGTM!

File: src/main/java/org/opensearch/timeseries/model/Config.java [hunk 6]
```diff
@@ -559,4 +562,14 @@ protected void checkAndThrowValidationErrors(ValidationAspect validationAspect)
             throw new TimeSeriesException(CommonMessages.FAIL_TO_VALIDATE);
         }
     }
+
+    public static Config parseConfig(Class<? extends Config> configClass, XContentParser parser) throws IOException {
+        if (configClass == AnomalyDetector.class) {
```
[the remark this answers is not in the document]
Comment (kaituo) [reply src/main/java/org/opensearch/timeseries/model/Config.java:567]: In order to apply the 'instanceof' operator, we require an existing instance of the object. However, in this case, the caller doesn't possess an instance yet. The primary objective of this method is, in fact, to generate such an instance.

File: src/main/java/org/opensearch/timeseries/model/Config.java (hunk 5 above)
Comment (kaituo) [reply src/main/java/org/opensearch/timeseries/model/Config.java:572]: Changed.

File: src/main/java/org/opensearch/timeseries/model/Config.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (owaiskazi19) [reply src/main/java/org/opensearch/timeseries/model/Config.java:567]: Can you use `equals` here instead of `==`?

**Comment by owaiskazi19** [comment]
Should we wait for nightly build for CI to pass?

File: src/main/java/org/opensearch/timeseries/model/Config.java (hunk 5 above)
Comment (owaiskazi19) [reply src/main/java/org/opensearch/timeseries/model/Config.java:572]: Will you be pushing all the changes together?

Comment (kaituo) [reply src/main/java/org/opensearch/timeseries/model/Config.java:572]: yes

**Comment by kaituo** [comment]
> Should we wait for updated nightly build for CI to pass?

We don't need to. The CI has been failing due to that for more than 1 week now.

File: src/main/java/org/opensearch/timeseries/model/Config.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (kaituo) [reply src/main/java/org/opensearch/timeseries/model/Config.java:567]: In Java, == is used for reference equality, meaning it checks whether two references point to the exact same object. When comparing classes, you want to know if configClass is exactly AnomalyDetector.class or Forecaster.class, not if it is an instance that is equal to those classes in some way.

On the other hand, equals() is used for logical equality and must be implemented by the class. It checks whether two objects should be considered "equal" based on their internal state, not their identity. However, equals() isn't meaningful when comparing Class objects, because there isn't a logical way to say two distinct classes are "equal."

So in summary, your current code is correct and there's no need to replace == with equals().

**Comment by kaituo** [comment]
Bwc test failed:

Execution failed for task ':adBwcCluster#mixedClusterTask'.

    cluster{::adBwcCluster0} failed to wait for cluster health yellow after 40 SECONDS
    IO error while waiting cluster
    503 Service Unavailable

Known issue: https://github.com/opensearch-project/OpenSearch/issues/5076
latest development: https://github.com/opensearch-project/OpenSearch/issues/5091

**Review by owaiskazi19 — APPROVED** [review]
